dali/pipeline/operators/arithmetic/expression_tree.h

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_TREE_H_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_TREE_H_
+
+#include "dali/core/any.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/util/backend2workspace_map.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+struct SampleExtent {
+  int sample_idx;
+  int extent_idx; // extents are number inside given sample_idx
+  int task_idx; // thread assgined to this extent
+  int64_t extent_size;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const SampleExtent& v) {
+  os << "{ " << v.sample_idx << ", " << v.extent_idx << ", " << v.task_idx << ", " << v.extent_size
+     << "}";
+  return os;
+}
+
+/**
+ * @brief Part responsible for loop-execution over tile, the implementation knows the types
+ * and can access the inputs internally
+ */
+class ExpressionImplBase {
+ public:
+  virtual void Execute(ArgumentWorkspace &ws, const OpSpec &spec, SampleExtent tile) = 0;
+  virtual ~ExpressionImplBase() = default;
+};
+
+/**
+ * @brief base class for nodes of ExpressionTree
+ *
+ * TODO(klecki): Using unique_ptrs for ownership, we may generalize it to DAG, that would require
+ * some changes
+ */
+struct ExpressionBase {
+
+  ExpressionBase() = default;
+
+  virtual const std::string &GetOp() const {
+    DALI_FAIL("No op in this expression node");
+  }
+  virtual bool IsLeaf() const {
+    return false;
+  }
+  virtual bool IsTensor() const {
+    return false;
+  }
+  virtual int GetMappedInput() const {
+    return -1;
+  }
+
+  DALIDataType type_id = DALI_NO_TYPE;
+  kernels::TensorListShape<> shape;
+  std::unique_ptr<ExpressionImplBase> impl;
+
+  virtual int GetSubexpressionCount() const {
+    return 0;
+  }
+
+  virtual void AddSubexpression(std::unique_ptr<ExpressionBase> expr) {
```

Comment:
This code smells of bad design - if it doesn't have an implementation, then it's not really a part of the base interface.

Reply:
Done